### PR TITLE
Add S3 key utils

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/6827

Problem Summary:
Split from https://github.com/pingcap/tiflash/pull/6874

### What is changed and how it works?

`S3Filename`: Generate the S3 object keys for local file or parsing the key from S3 object key
`TiFlashS3Client`: Aws::S3::S3Client with bucket to simplify some argument passing
`DB::S3::listPrefix`: handle the `LIST` result is more than 1000 by passing a callback function

https://github.com/pingcap/tiflash/blob/e323ceb3ce34fdf1c3ff69e2871e4d2c123bc9ad/dbms/src/Storages/S3/S3Filename.h#L43-L68

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/S3/S3Filename.h [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+#include <fmt/format.h>
+
+#include <string_view>
+
+namespace DB::S3
+{
+
+struct DMFileOID
+{
+    UInt64 write_node_id = 0;
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply dbms/src/Storages/S3/S3Filename.h:28]: yes

File: dbms/src/Storages/S3/S3Common.h [hunk 2]
```diff
@@ -71,7 +93,34 @@ bool objectExists(const Aws::S3::S3Client & client, const String & bucket, const
...
+
+
+std::pair<bool, Aws::Utils::DateTime> tryGetObjectModifiedTime(
+    const Aws::S3::S3Client & client,
+    const String & bucket,
+    const String & key);
+
+void deleteObject(const Aws::S3::S3Client & client, const String & bucket, const String & key);
+
 std::unordered_map<String, size_t> listPrefix(const Aws::S3::S3Client & client, const String & bucket, const String & prefix);
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply dbms/src/Storages/S3/S3Common.h]: removed

**Review by lidezhu — APPROVED** [review]
LGTM

**Pull request merged**